### PR TITLE
bpo-46301: cover uncomparable values in `Enum._convert_`

### PR DESCRIPTION
**Pull request by sobolevn**
https://bugs.python.org/issue46301

### CODE REVIEW

File: Lib/test/test_enum.py [hunk 1]
```diff
@@ -4477,6 +4485,19 @@ def test_convert(self):
                           and name not in dir(IntEnum)],
                          [], msg='Names other than CONVERT_TEST_* found.')
 
+    def test_convert_uncomparable(self):
+        uncomp = enum.Enum._convert_(
```
Comment (sobolevn) [inline Lib/test/test_enum.py:4490]: Moreover, `_convert_` was not tested on plain `Enum`

**Comment by sobolevn** [comment]
@AlexWaygood can you please add "skip news" to this PR? I can't add labels myself 🙂

File: Lib/test/test_enum.py (hunk 1 above)
Comment (ethanfurman) [reply Lib/test/test_enum.py:4490]: `_convert_` was not intended to be called on plain `Enum` -- it's purpose it to convert existing constants in a module to enums, and existing constants will already be `str` or `int` (or possibly something else, like `complex`).

File: Lib/test/test_enum.py [hunk 2]
```diff
@@ -4440,6 +4440,14 @@ def test__all__(self):
 CONVERT_STRING_TEST_NAME_E = 5
 CONVERT_STRING_TEST_NAME_F = 5
 
+# We also need values that cannot be compared:
```
Comment (ethanfurman) [inline Lib/test/test_enum.py:4443]: All we need for this test is for `UNCOMPARABLE_*` to be different types of values:

`UNCOMPABLE_A` = 5
`UNCOMPABLE_C` = (9, 4)
`UNCOMPABLE_B` = 'hello'

File: Lib/test/test_enum.py (hunk 1 above)
Comment (ethanfurman) [reply Lib/test/test_enum.py:4490]: Change the `UNCOMPARABLE_*` values as suggested above, and also add a test using a `complex` type.

**Review by ethanfurman — CHANGES_REQUESTED** [review]

**Comment by sobolevn** [comment]
Thanks! I've added one more case with `complex` and simplified `UNCOMPARABLE_` defs.

I have made the requested changes; please review again

**Review by ethanfurman — APPROVED** [review]

**Comment by sobolevn** [comment]
Thank you! 👍

**Comment by pablogsal** [comment]
Seems that this PR has broken the refleak buildbots:

https://buildbot.python.org/all/#/builders/320/builds/269/
https://buildbot.python.org/all/#/builders/384/builds/255

According to our buildbot policy, if this is not fixed in 24h we will need to revert

**Comment by sobolevn** [comment]
@pablogsal thanks for the info! I would love to help to get this fixed, but I am not familiar with how refleak works. And since this is a test-only PR I am bit surprised that this happened 🤔

**Comment by sobolevn** [comment]
I was able to reproduce this locally on macos with `./python.exe -m test -R : -v test_enum`

```
test_enum leaked [220, 220, 220, 220] references, sum=880
test_enum leaked [56, 56, 56, 56] memory blocks, sum=224
test_enum failed (reference leak) in 39.1 sec

== Tests result: FAILURE ==

1 test failed:
    test_enum

Total duration: 39.1 sec
Tests result: FAILURE
```

**Comment by sobolevn** [comment]
`PYTHONDUMPREFSFILE=ex.txt ./python.exe -m test -R : -v test_enum` produces this:
[ex.txt](https://github.com/python/cpython/files/7836174/ex.txt)

**Comment by pablogsal** [comment]
I am not at my computer so I cannot test, but the problem that happens in these cases where only tests are added is that the new tests are either modifying the global state or adding some sort of cache.

**Comment by sobolevn** [comment]
Ok, looks like I've found the problem: https://github.com/python/cpython/blob/1bee9a4625e101d3308831de37590f4e2f57c71c/Lib/enum.py#L833

When I comment this line out - tests pass, with it - we have a refleak.
@

**Comment by pablogsal** [comment]
> Ok, looks like I've found the problem:

Yeah, this is an example of "modifying the global state". Tests should be idempotent, and they should leave everything as it was before

**Comment by sobolevn** [comment]
> I am not at my computer so I cannot test, but the problem that happens in these cases where only tests are added is that the new tests are either modifying the global state or adding some sort of cache.

Yes, you are right! These tests do modify the global state. They add a new class to the current module.

> Yeah, this is an example of "modifying the global state". Tests should be idempotent, and they should leave everything as it was before

But this line is from `enum.py`, not from tests 🤔 

Either way, I will open a new PR with `test-with-buildbots` later today. I hope that I will be able to fix it!

**Comment by ethanfurman** [comment]
The line `body['__module__'] = module` is not modifying global state: `body` is a dict that will be used to create a new enum class.

The clue was good, though -- the global constants that are converted into enums need to be reset to their original values, either before or after the test runs.

**Comment by ethanfurman** [comment]
@sobolevn, add

        global COMPLEX_A, COMPLEX_B, COMPLEX_C

to the top of `test_convert_complex`, and

        COMPLEX_A = 2j
        COMPLEX_B = 3j
        COMPLEX_C = 1j

to the bottom.  Make similar changes to `test_convert_uncomparable`.

**Comment by sobolevn** [comment]
Oh, sorry, I've missed your comment in the morning 🙂 
I went with `.swap_item` technique. This ensures that we don't polute real modules.

**Pull request merged**